0001-Enable-pbuffer-to-fix-most-deqp-pbuffer-testcases.patch

```diff
@@ -0,0 +1,291 @@
+
+diff --git a/src/egl/drivers/dri2/egl_dri2.h b/src/egl/drivers/dri2/egl_dri2.h
+index 3add32e..f3d09dc 100644
+--- a/src/egl/drivers/dri2/egl_dri2.h
++++ b/src/egl/drivers/dri2/egl_dri2.h
+@@ -290,7 +290,8 @@ struct dri2_egl_surface
+ #ifdef HAVE_ANDROID_PLATFORM
+    struct ANativeWindow *window;
+    struct ANativeWindowBuffer *buffer;
+-   __DRIimage *dri_image;
++   __DRIimage *dri_image_back;
++   __DRIimage *dri_image_front;
+ 
+    /* EGL-owned buffers */
+    __DRIbuffer           *local_buffers[__DRI_BUFFER_COUNT];
+diff --git a/src/egl/drivers/dri2/platform_android.c b/src/egl/drivers/dri2/platform_android.c
+index ec52a02..6b4f22f 100644
+--- a/src/egl/drivers/dri2/platform_android.c
++++ b/src/egl/drivers/dri2/platform_android.c
+@@ -204,9 +204,9 @@ droid_window_enqueue_buffer(_EGLDisplay *disp, struct dri2_egl_surface *dri2_sur
+ 
+    mtx_lock(&disp->Mutex);
+ 
+-   if (dri2_surf->dri_image) {
+-      dri2_dpy->image->destroyImage(dri2_surf->dri_image);
+-      dri2_surf->dri_image = NULL;
++   if (dri2_surf->dri_image_back) {
++      dri2_dpy->image->destroyImage(dri2_surf->dri_image_back);
++      dri2_surf->dri_image_back = NULL;
+    }
+ 
+    return EGL_TRUE;
+@@ -295,7 +295,7 @@ droid_create_surface(_EGLDriver *drv, _EGLDisplay *disp, EGLint type,
+       window->query(window, NATIVE_WINDOW_HEIGHT, &dri2_surf->base.Height);
+    }
+ 
+-   config = dri2_get_dri_config(dri2_conf, EGL_WINDOW_BIT,
++   config = dri2_get_dri_config(dri2_conf, type,
+                                 dri2_surf->base.GLColorspace);
+    if (!config)
+       goto cleanup_surface;
+@@ -353,6 +353,18 @@ droid_destroy_surface(_EGLDriver *drv, _EGLDisplay *disp, _EGLSurface *surf)
+       dri2_surf->window->common.decRef(&dri2_surf->window->common);
+    }
+ 
++   if (dri2_surf->dri_image_back) {
++       _eglLog(_EGL_DEBUG, "%s : %d : destroy dri_image_back", __func__, __LINE__);
++      dri2_dpy->image->destroyImage(dri2_surf->dri_image_back);
++      dri2_surf->dri_image_back = NULL;
++   }
++
++   if (dri2_surf->dri_image_front) {
++       _eglLog(_EGL_DEBUG, "%s : %d : destroy dri_image_front", __func__, __LINE__);
++      dri2_dpy->image->destroyImage(dri2_surf->dri_image_front);
++      dri2_surf->dri_image_front = NULL;
++   }
++
+    (*dri2_dpy->core->destroyDrawable)(dri2_surf->dri_drawable);
+ 
+    free(dri2_surf);
+@@ -384,50 +396,122 @@ update_buffers(struct dri2_egl_surface *dri2_surf)
+ }
+ 
+ static int
+-get_back_bo(struct dri2_egl_surface *dri2_surf)
++droid_get_front_bo(struct dri2_egl_surface *dri2_surf, unsigned int format)
+ {
+    struct dri2_egl_display *dri2_dpy =
+       dri2_egl_display(dri2_surf->base.Resource.Display);
+-   int fourcc, pitch;
+-   int offset = 0, fd;
+ 
+-   if (dri2_surf->dri_image)
+-	   return 0;
++   if (dri2_surf->dri_image_front)
++   {
++      _eglLog(_EGL_WARNING, "%s:%d dri2_image_front allocated !\n", __func__, __LINE__);
++      return 0;
++   }
+ 
```

Comment:
Please hold a while, as this patch will expose another issue in current mesa window surface implementation. Screen will flick sometimes,  as Android UI uses partial update,  but mesa window  surface implementation does not support preserve swap behavior.  
We can hold or merge this patch in Android-IA,  Zhiquan will make another WA for window surface, then send the new patch set to upstream mail list